### DISCUSSION TITLE
ST7735 Display - ESP32

### DISCUSSION THREAD

**Discussion by 4-digital** [arendst/Tasmota#19080] [Q&A] [answered]
I'm trying to connect an ST7735 display, I see that it's supported in version 13 but I'm having trouble finding the necessary information. Which option should I choose for the DisplayModel and what pins should I use for an ESP32? These are the pins from the display: 
1 GND
2 VCC
3 SCK
4 SDA
5 RES
6 RS
7 CS
8 LED
Thank you for your help!

**Comment by sfromis** [comment]
The preferred way is not by DisplayModel, but instead the [Universal Display Driver](https://tasmota.github.io/docs/Displays/#universal-display-driver), and a display descriptor to match your display. Available ones does include ST7735. That docs section also alludes to the pins to be configured in Tasmota. Naming on the board can be a bit confusing, but the product docs should explain what's what.

**Comment by 4-digital** [comment]
@sfromis, thank you for your response. I thought that since it is supported by version 13, it gives me the option to configure it in GUI but that's not possible. I tried the option you indicated but still without result. I'm not sure I followed the correct steps: I compiled the tasmota32-display with the #define USE_UNIVERSAL_DISPLAY option. I saved the contents of the ST7735_128x160_display.ini file in the display.ini file, modifying the first line with the pins I'm using. I uploaded this file to the ESP32 through the Manage File System. I configured a GPIO pin with OptionA3. From the console I set DisplayModel 17, DisplayMode 1. I also tried the option to leave the character * in the display.ini file and define the pins used through the graphical interface. I'm not a programmer and a link with more detailed steps on how to use the Universal Display Driver would help me. Thank you.

**Reply by sfromis** [reply, marked as answer]
You should not need to do anything with DisplayModel, it should become 17 automatically if you have the right setup. What exact gpio config did you do? And wired to what? 

What I'd suggest would be to use the precompiled tasmota32-lvgl, which has all the uDisplay support already.

IMO, it is preferable to keep pins as * in the display.ini file, and define all pins in a template/module config.

**Reply by sfromis** [reply]
When booting with a working setup, you should get messages similar to these:
``` 
00:00:00.038 SPI: Hardware using GPIO07(CLK), GPIO11(MOSI) and GPIO09(MISO)
00:00:00.549 DSP: Configured display 'ILI9341'
```

**Reply by martin2705** [reply]
@ 4-digital: Can you give me an advise how you connect to the display? I tried a lot but do not find the right connection.
I also got the message with Hardware using Pins, but than there is an error: 00:00:00.064 TFS: File 'mcp23x.dat' not found
So I think, there are some mistakes in the pinning.
my connections:
1 GND - GND
2 VCC - VCC
3 SCK - SPI-CLK
4 SDA - SPI-MOSI
5 RES- VCC
6 RS - GND
7 CS - SPI-CS
8 LED - VCC
Thanks

**Reply by sfromis** [reply]
`00:00:00.064 TFS: File 'mcp23x.dat' not found` is not an error message, just an information message rarely relevant. Other messages around that time are more interesting.

But a key point is also what exact binary you flashed with, like if it supports the display. Generally recommended is, as I pointed out earlier, to use the universal display driver with a `-lvgl` binary variant, making sure to config `Option A 3` and having a display descriptor file.

**Reply by martin2705** [reply]
Hi,
I flashed as suggested the precompiled tasmota32-lvgl

Nothing special around:
00:00:00.003 HDW: ESP32-D0WD-V3 v3.0 
00:00:00.040 UFS: FlashFS mounted with 304 kB free
00:00:00.062 CFG: Loaded from File, Count 28
00:00:00.063 FRC: Some settings have been reset (2)
00:00:00.067 SPI: Hardware using GPIO18(CLK), GPIO23(MOSI) and GPIO19(MISO)
00:00:00.072 TFS: File 'mcp23x.dat' not found
00:00:00.154 BRY: Berry initialized, RAM used 5668 bytes
00:00:00.534 DSP: Configured display 'ST7735'
00:00:00.549 Project tasmota - Tasmota Version 13.3.0(lvgl-haspmota)-2_0_14(2023-12-12T14:31:28)
00:00:00.557 LVG: LVGL initialized

Option A 3 is on IO4
Display.ini contains ST7735_128x160_display.ini

Now I get a messagte that the display is configured, there was a mistake in the display.ini.
But the display shows nothing, only blank. :(

**Reply by sfromis** [reply]
Did you also turn on the display? Power command or front page toggle. 

When it comes to what's in the log, seeing the messages about SPI, DSP and LVG are great to confirm those important initialization steps.

However, there is very much "something special around"; the message `FRC: Some settings have been reset (2)` is a bad sign of trouble, meaning that Tasmota was detecting a boot loop (quickly crashing after boot) which can lead to Tasmota starting the "scale back" the config, disabling things until able to boot without crashing. However, "only" getting to 2 does mean that it was not "too persistent".

One risk factor with a display is if you're not supplying the board with sufficient power. USB ports on a computer are often rather weak sources of power.

**Reply by martin2705** [reply]
Got it. You have to connect every Pin,
1 GND - GND
2 VCC - VCC
3 SCK - SPI-CLK
4 SDA - SPI-MOSI
5 RES- VCC has to be Display Rst
6 RS - GND has to be DC
7 CS - SPI-CS
8 LED - has to be Backlight
Now that was too simple.
With the ideas to set some Pins to GND or VCC my display do no work anyway.

**Reply by moodyblue** [reply]
I'm trying to use a ESP32C3 device that has this diagram
[ESP32-C3-1.44inch.pdf](https://github.com/user-attachments/files/21663578/ESP32-C3-1.44inch.pdf)
I've flashed tasmota32c3-lvgl.factory.bin
Option A 3 is on IO4
Display.ini contains ST7735_128x160_display.ini
Here is the log
```
00:00:00.200 CMD: Using USB CDC
00:00:00.201 HDW: ESP32-C3 v0.4 
00:00:00.215 UFS: FlashFS mounted with 12592 kB free
00:00:00.224 CFG: Loaded from File, Count 44
00:00:00.227 SER: Set to 8N1 115200 bit/s
00:00:00.227 SER: HWCDC supports 115200 bit/s only
00:00:00.232 QPC: Count 1
00:00:00.276 BRY: Berry initialized, RAM used 5572 bytes
00:00:00.673 DSP: ST7735 initialized
00:00:00.683 Project tasmota - Tasmota Version 15.0.1.3(5bd9414-lvgl-haspmota)-3_1_3(2025-08-04T10:12:12)
00:00:00.689 RSL: RESULT = {"POWER":"ON"}
00:00:00.690 RSL: POWER = ON
00:00:00.698 LVG: LVGL initialized
00:00:02.001 WIF: Connecting to AP1 Aveiro-Domotic Channel 6 BSSId F6:F2:6D:2C:BB:5B in mode HT40 as displaygaragem-5108...
00:00:03.997 WIF: Connected
00:00:04.254 HTP: Web server active on displaygaragem-5108 with IP address 192.168.129.70
00:00:06.885 MQT: Attempting connection...
00:00:07.052 MQT: Connected
00:00:07.055 MQT: tele/displaygaragem/LWT = Online (retained)
00:00:07.057 MQT: cmnd/displaygaragem/POWER = 
00:00:07.061 MQT: tele/displaygaragem/INFO1 = {"Info1":{"Module":"ESP32C3","Version":"15.0.1.3(5bd9414-lvgl-haspmota)","FallbackTopic":"cmnd/DVES_E4D3F4_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:07.064 MQT: tele/displaygaragem/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"displaygaragem-5108","IPAddress":"192.168.129.70","IP6Global":"","IP6Local":"fe80::26ec:4aff:fee4:d3f4%st1"}}
00:00:07.067 MQT: tele/displaygaragem/INFO3 = {"Info3":{"RestartReason":"Vbat power on reset","BootCount":16}}
00:00:07.070 MQT: stat/displaygaragem/RESULT = {"POWER":"ON"}
00:00:07.071 MQT: stat/displaygaragem/POWER = ON
00:00:07.081 QPC: Reset
12:06:11.227 CMD: displaymodel
12:06:11.231 MQT: stat/displaygaragem/RESULT = {"DisplayModel":17}
12:06:19.618 CMD: displaymode
12:06:19.623 MQT: stat/displaygaragem/RESULT = {"DisplayMode":0}
12:07:32.976 CMD: displaytext qwer
12:07:32.985 MQT: stat/displaygaragem/RESULT = {"DisplayText":"qwer"}
```
But the display is blank. Probably a pin error in the display.ini file but I don't know how to correct it.

**Reply by sfromis** [reply]
I'm assuming that you are talking about this specific `display.ini`:
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7735_128x160_display.ini

As there are no pins defined in that display descriptor, there can be no "pin error" in that file. Configuring the SPI pins is up to you. What's your configuration of those pins?

Also, it seems odd that you placed `Option A 3` on gpio 4, when the diagram you posted suggests that gpio 4 is connected to the SPI MOSI pin.

**Reply by moodyblue** [reply]
Thanks for the quick answer. I've corrected the A3 assignment and now the module parameters are:
<img width="423" height="769" alt="imagem" src="https://github.com/user-attachments/assets/bc936428-caa5-48c7-81ca-b067df27f165" />

**Reply by sfromis** [reply]
Well, all the LCD pins (gpio 0, 2-5) from the diagram are relevant, including SCK=CLK. Probably, LCD_RST is config `Display Rst`.

**Reply by moodyblue** [reply]
> Well, all the LCD pins (gpio 0, 2-5) from the diagram are relevant, including SCK=CLK. Probably, LCD_RST is config `Display Rst`.

You are the greatest. Now the device boots and briefly displays a Tasmota logo. Looks like 128x160 is not correct for this screen so I changed display.ini for 128x128. Seems better now.

My next task is to discover how to display messages received through mqtt.

**Reply by moodyblue** [reply]
I've now completed my project. It's documented here https://community.openhab.org/t/esp32-with-a-small-display/165391/1

Kudos to @sfromis for his decisive help
